docs/tracing.md

```diff
@@ -0,0 +1,246 @@
+
+It is disabled by default at compile-time, meaning the tracing
+functionality has no impact on the release binary.
+
+To enable tracing in Firecracker, compile with `--features tracing` e.g.
+`cargo build --features tracing`.
+
+This will result in an increase in the binary size (~100kb) and a
+significant regression in performance (>10x). To mitigate the
+performance impact you can filter the tracing output as described in the
+next section.
+
+## Filtering
+
+You can filter tracing output both at run-time and compile-time.
+This can be used to mitigate the performance impact of logging many
+traces.
+
+Run-time filtering is implemented with the `/logger` API call, this can
+significantly mitigate the impact on execution time but cannot mitigate
+the impact on memory usage.
+
+Compile-time filtering is a manual process assisted by the
+[`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)
+tool, this can almost entirely mitigate the impact on execution time and
+the impact on memory usage.
+
+### Run-time
+
+You can filter by module path and/or file path at runtime, e.g.:
+
+```bash
+curl -X PUT --unix-socket "${API_SOCKET}" \
+    --data "{
+        \"level\": \"Trace\",
+        \"module\": \"api_server::request\",
+    }" \
+    "http://localhost/logger"
+```
+
+Instrumentation logs are `Trace` level logs, at runtime the level must
+be set to `Trace` to see them. The module filter applied here ensures
+only logs from the `request` modules within the `api_server` crate will
+be output.
+
+This will mitigate most of the performance regression.
+
+### Compile-time
+
+Specific environments can restrict run-time configuration, in these
+environments it becomes necessary to support targeted tracing without
+run-time re-configuration, for this compile-time filtering must be used.
+
+To reproduce the same filtering as run-time at compile-time, you can use
+[`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)
+at compile-time like:
+
+```bash
+# Install clippy-tracing
+cargo install clippy-tracing --features log
+# Remove all instrumentation.
+clippy-tracing --action strip --path ./src
+# Adds instrumentation to the specific file/s.
+clippy-tracing --action fix --path ./src/api_server/src/request --cfg-attr "feature = \"tracing\""
+# Build Firecracker
+cargo build --features tracing
+```
+
+Then at run-time:
+
+```bash
+curl -X PUT --unix-socket "${API_SOCKET}" \
+    --data "{
+        \"level\": \"Trace\",
+    }" \
+    "http://localhost/logger"
+```
+
+The instrumentation has been stripped from all files other than those at
+`./src/api_server/src/request` so we do not need to apply a run-time
+filter.
+
+## Example
+
+In this example we start Firecracker with tracing then make a simple API
+call.
+
+### API call
+
+```
+jcawl@u08fff4fc3ef156:~/Projects/firecracker$ sudo curl -X GET --unix-socket "/run/firecracker.socket" "http://localhost/"
```

Comment:
nit: here and below: can anonymise the prompt.

Reply:
I've removed `jcawl@u08fff4fc3ef156:`